File: Python/Construct Binary Tree from Preorder and Inorder Traversal.py
"""
Given preorder and inorder traversal of a tree, construct the binary tree.
"""

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param preorder, a list of integers
    # @param inorder, a list of integers
    # @return a tree node
    def buildTree(self, preorder, inorder):
        if len(inorder) == 0:
            return None
        root = TreeNode(preorder.pop(0))
        stn = list()
        stn.append(root)

        while True:
            if inorder[0] == stn[-1].val:
                p = stn.pop()
                inorder.pop(0)
                if len(inorder) == 0: break
                if len(stn) != 0 and inorder[0] == stn[-1].val: continue
                p.right = TreeNode(preorder.pop(0))
                stn.append(p.right)
            else:
                p = TreeNode(preorder.pop(0))
                stn[-1].left = p
                stn.append(p)

        return root

File: Java/Construct Binary Tree from Preorder and Inorder Traversal.java
/*
Given preorder and inorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(inorder.length == 0) return null;
        
        TreeNode root = new TreeNode(preorder[0]);
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        
        int idx1 = 0;
        int idx2 = 1;
        while(true){
            if(inorder[idx1] == stack.peek().val){
                TreeNode p = stack.pop();
                idx1++;
                if(idx1 == inorder.length) break;
                if(!stack.isEmpty() && stack.peek().val == inorder[idx1]) continue;
                p.right = new TreeNode(preorder[idx2]);
                idx2++;
                stack.push(p.right);
            }
            else{
                TreeNode p = new TreeNode(preorder[idx2]);
                idx2++;
                stack.peek().left = p;
                stack.push(p);
            }
        }
        
        return root;
    }
}
